### PR TITLE
Repetition penalty

### PR DESCRIPTION
**Pull request by paudelnirajan**
* repetition penalty applied to both draft and target logits.

* @look-luc mentioned sth called thinking tokens being shown in the response, qwen3 models generate reasoning tokens so, i have filtered those out. I have standalone run story file if you want to test. i think @look-luc also has a PR relating this. He mentioned that problem so, i tried this.

### CODE REVIEW

**Review by look-luc — COMMENTED** [review]
Looks good to me

**Review by look-luc — APPROVED** [review]

File: src/spec_decode.py [hunk 1]
```diff
@@ -259,18 +309,37 @@ def get_time():
...
+            n_draft = new_draft_tokens.size(-1)  # may be < gamma if EOS was hit early
+            target_raw_logits = target_out.logits  # (bs, n_draft+1, d_vocab)
+            penalized_target_logits = torch.empty_like(target_raw_logits)
+            for j in range(n_draft + 1):
+                # draft_so_far at verification position j equals the first j draft tokens
+                penalized_target_logits[:, j, :] = penalize_logits(
+                    target_raw_logits[:, j, :],
+                    confirmed_len=cur_gen_idx,
+                    draft_so_far=new_draft_tokens[:, :j],
+                )
```
[the remark this answers is not in the document]
Comment (michaelpginn) [reply src/spec_decode.py]: I think this is correct and important

**Comment by michaelpginn** [comment]
Have we tested whether this actually eliminates the repetition issue?

File: src/spec_decode.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (paudelnirajan) [reply src/spec_decode.py]: i tried batching the verifier pass, that should reduce some time overhead?

File: src/spec_decode.py [hunk 2]
```diff
@@ -99,6 +99,8 @@ def speculative_decode(
     gamma: int = 5,
     top_k: int = 0,
     top_p: float = 0.0,
+    repetition_penalty: float = 1.0,
+    repetition_penalty_window: int = 64,
```
Comment (michaelpginn) [inline src/spec_decode.py]: this seems like a pretty big window, is this typical?

Comment (paudelnirajan) [reply src/spec_decode.py]: no i was trying various ones --  i tested with window of 16 for the graphs i shared in group

that is from my first test which was not good

**Review by michaelpginn — CHANGES_REQUESTED** [review]
Just merging the story gen stuff

File: src/spec_decode.py [hunk 3]
```diff
@@ -467,3 +532,67 @@ def apply_top_p(logits: torch.Tensor, p: float) -> torch.Tensor:
...
+    logits: torch.Tensor,
+    context_ids: torch.Tensor,
+    penalty: float,
+) -> torch.Tensor:
+    """Apply multiplicative repetition penalty to raw logits (single position)."""
+    if penalty == 1.0 or context_ids.size(-1) == 0:
+        return logits
+
+    for b in range(logits.size(0)):
+        window_counts = torch.nn.functional.one_hot(context_ids[b]).sum(dim=1)
```
Comment (paudelnirajan) [inline src/spec_decode.py]: hey @michaelpginn , do you think -- here `dim` should be `0`?

Comment (michaelpginn) [reply src/spec_decode.py]: Yes good catch. We could also make both of them dim=-2 so it's consistent?

File: src/story_gen_prompt.py
Comment (michaelpginn) [inline src/story_gen_prompt.py:1]: Can we keep this stuff? I think we'd rather generate stories from a short phrase versus a longer monolingual text?

Comment (michaelpginn) [reply src/story_gen_prompt.py:1]: The `load_data` function above could just use this approach to make a bunch of adjective/noun combos.

Comment (paudelnirajan) [reply src/story_gen_prompt.py:1]: so the combos(story topic) will be english ? and we will ask in prompt to generate story in target language ?

Comment (michaelpginn) [reply src/story_gen_prompt.py:1]: That was my thinking, unless that's making it less likely to generate text in the right language

Comment (paudelnirajan) [reply src/story_gen_prompt.py:1]: it's fine i think. it's similar to translation stuff we are doing. `input_ids` in english and decoding in `target_lang`. i was just making sure!

File: src/spec_decode.py (hunk 3 above)
Comment (paudelnirajan) [reply src/spec_decode.py]: yes done!

**Review by michaelpginn — APPROVED** [review]
